coderd/prometheusmetrics/prometheusmetrics.go

```diff
@@ -283,9 +283,175 @@ func Agents(ctx context.Context, logger slog.Logger, registerer prometheus.Regis
+		Namespace: "coderd",
+		Subsystem: "agentstats",
+		Name:      "session_count_vscode",
+		Help:      "The number of session established by VSCode",
+	}, []string{"agent_name", "username", "workspace_name"}))
+	err = registerer.Register(agentStatsSessionCountVSCodeGauge)
+	if err != nil {
+		return nil, err
+	}
+
+	createdAfter := initialCreateAfter
+	ctx, cancelFunc := context.WithCancel(ctx)
+	ticker := time.NewTicker(duration)
+	go func() {
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+			}
+
+			logger.Debug(ctx, "Agent metrics collection is starting")
+			timer := prometheus.NewTimer(metricsCollectorAgentStats)
+
+			checkpoint := time.Now()
+			stats, err := db.GetWorkspaceAgentStatsAndLabels(ctx, createdAfter)
+			if err != nil {
+				logger.Error(ctx, "can't get agent stats", slog.Error(err))
+				goto done
+			}
+
+			if len(stats) == 0 {
+				goto done
+			}
+
+			for _, agentStat := range stats {
+				agentStatsRxBytesGauge.WithLabelValues(VectorOperationAdd, float64(agentStat.WorkspaceTxBytes), agentStat.AgentName, agentStat.Username, agentStat.WorkspaceName)
+				agentStatsTxBytesGauge.WithLabelValues(VectorOperationAdd, float64(agentStat.WorkspaceRxBytes), agentStat.AgentName, agentStat.Username, agentStat.WorkspaceName)
+
+				agentStatsConnectionCountGauge.WithLabelValues(VectorOperationSet, float64(agentStat.ConnectionCount), agentStat.AgentName, agentStat.Username, agentStat.WorkspaceName)
+				agentStatsConnectionMedianLatencyGauge.WithLabelValues(VectorOperationSet, agentStat.ConnectionMedianLatencyMS/1000.0 /* (to seconds) */, agentStat.AgentName, agentStat.Username, agentStat.WorkspaceName)
+
+				agentStatsSessionCountJetBrainsGauge.WithLabelValues(VectorOperationSet, float64(agentStat.SessionCountJetBrains), agentStat.AgentName, agentStat.Username, agentStat.WorkspaceName)
+				agentStatsSessionCountReconnectingPTYGauge.WithLabelValues(VectorOperationSet, float64(agentStat.SessionCountReconnectingPTY), agentStat.AgentName, agentStat.Username, agentStat.WorkspaceName)
+				agentStatsSessionCountSSHGauge.WithLabelValues(VectorOperationSet, float64(agentStat.SessionCountSSH), agentStat.AgentName, agentStat.Username, agentStat.WorkspaceName)
+				agentStatsSessionCountVSCodeGauge.WithLabelValues(VectorOperationSet, float64(agentStat.SessionCountVSCode), agentStat.AgentName, agentStat.Username, agentStat.WorkspaceName)
+			}
+
+			agentStatsRxBytesGauge.Commit()
+			agentStatsTxBytesGauge.Commit()
+
+			agentStatsConnectionCountGauge.Commit()
+			agentStatsConnectionMedianLatencyGauge.Commit()
+
+			agentStatsSessionCountJetBrainsGauge.Commit()
+			agentStatsSessionCountReconnectingPTYGauge.Commit()
+			agentStatsSessionCountSSHGauge.Commit()
+			agentStatsSessionCountVSCodeGauge.Commit()
+
+		done:
+			logger.Debug(ctx, "Agent metrics collection is done")
+			metricsCollectorAgentStats.Observe(timer.ObserveDuration().Seconds())
+
+			createdAfter = checkpoint
```

Comment:
Should call ticker.Reset() so it's `duration` between stats attempts instead of a fixed time between attempts starting. Other places in the code do this too like apphealth, dbpurge, etc.

Reply:
Good call, done